tests/testsuite/deduplicate_workspace.rs

```diff
@@ -0,0 +1,806 @@
+//! Tests for deduplicating Cargo.toml fields with { workspace = true }
+use cargo_test_support::registry::{Dependency, Package};
+use cargo_test_support::{
+    basic_lib_manifest, basic_manifest, basic_workspace_manifest, git, path2url, paths, project,
+    publish, registry,
+};
+
+#[cargo_test]
+fn permit_additional_workspace_fields() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = ["bar"]
+            version = "1.2.3"
+            authors = ["Rustaceans"]
+            description = "This is a crate"
+            documentation = "https://www.rust-lang.org/learn"
+            readme = "README.md"
+            homepage = "https://www.rust-lang.org"
+            repository = "https://github.com/example/example"
+            license = "MIT"
+            license-file = "./LICENSE"
+            keywords = ["cli"]
+            categories = ["development-tools"]
+            publish = false
+            edition = "2018"
+
+            [workspace.badges]
+            gitlab = { repository = "https://gitlab.com/rust-lang/rust", branch = "master" }
+
+            [workspace.dependencies]
+            dep = "0.1"
+        "#,
+        )
+        .file("bar/Cargo.toml", &basic_workspace_manifest("bar", ".."))
+        .file("bar/src/main.rs", "fn main() {}")
+        .build();
+
+    p.cargo("build")
+        // Should not warn about unused fields.
+        .with_stderr(
+            "\
+[COMPILING] bar v0.1.0 ([CWD]/bar)
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+",
+        )
+        .run();
+
+    p.cargo("check").run();
+    let lockfile = p.read_lockfile();
+    assert!(!lockfile.contains("dep"));
```

Comment:
This should be failing.

I could use some guidance around updating `Cargo.lock`, as it appears that `[workspace.dependencies]` doesn't affect it unless they're referenced by a member. [When I tried to update this logic,](https://github.com/yaymukund/cargo/commit/3d6fd762122291cd4ee142d4f8ee237d684eaa0e) i ended up always compiling all workspace dependencies into all members, which is obviously wrong. Do we need a new `DepType`?